### ISSUE TITLE
pnpm: Package updated to wrong version

### ISSUE THREAD

**Issue by itpropro** [ScoopInstaller/Main#3663]
The GitHub actions pulls new releases without correctly comparing them if they are newer than the current version. 
E.g. pnpm is already at major version 7, but the GitHub action has still updated the current pnpm package from 7.19 to 6.32.21 (https://github.com/ScoopInstaller/Main/commit/bba8e4eb34310135bf700ed377d353dffa0ba86c), since it is the newest published package (but for the no longer maintained 6.x version).
Maybe it's already enough to change this line
https://github.com/ScoopInstaller/Scoop/blob/aaa726c09f786d2af902d827ae8d0caee964ca1e/bin/checkver.ps1#L324
to
`$update_available = (Compare-Version -ReferenceVersion $ver -DifferenceVersion $expected_ver) -eq 1`
to fix this.

**Comment by rasa** [comment]
https://github.com/pnpm/pnpm/releases shows [v6.32.21](https://github.com/pnpm/pnpm/releases/tag/v6.32.21) as the "Latest" version, and an update to [v7.1.9](https://github.com/pnpm/pnpm/releases/tag/v7.1.9). This isn't a scoop bug per se.

As a workaround, you can install v7.1.9 using
```sh
scoop install pnpm@7.1.9
```

**Comment by rashil2000** [comment]
> Handling the "lastest" version on GitHub as the always most current is definitely wrong

This is by design, it's not a bug.

However in pnpm's case though, the checkver can be configured to only look for 7.x.x releases. Can you make a PR for the same?

**Comment by itpropro** [comment]
Ok I understand, Scoop doesn't update to the latest version by design, not because it's obvious wrong behavior (which is always the case when relying on GitHub version numbers only).
Then please state this in the documentation, as many teams could then stop to rely on it if they want a stable production environment and use a package manager that handles update in the correct way.

**Comment by itpropro** [comment]
> 
> However in pnpm's case though, the checkver can be configured to only look for 7.x.x releases. Can you make a PR for the same?

I don't think this fixes the problem, it would only fix a symptom. The correct way to fix this would be to fix the behavior, how Scoop handles version updates (especially from GitHub) and how it defines if one version is newer than another. This should rely on semver wherever possible, which it doesn't do currently.
Could you also please reopen the issue and ask next time if the issue creator thinks his problem is solved?

**Comment by rashil2000** [comment]
Not all packages follow semver. It's very rare for package devs to release conflicting versions of packages in tandem (pnpm's the only case where I've seen this happen). So for an overwhelming majority of packages the latest GitHub release just works. 

Nevertheless, this is not the reason why Scoop will always update to latest release it finds. Scoop changes package versions depending on what it sees as latest, regardless of whether the change is an upgrade or a downgrade. This is because many times package devs recall latest releases due to critical bugs or security issues, and Scoop needs to keep up with that. 

pnpm is a rare case here and will need special-casing (for 7.x.x strings), as I said earlier.

**Comment by itpropro** [comment]
The regular way of recalling a package would be to create a bugfix or hotfix, which increases semver, e.g. going from 7.4.1 to 7.4.2.
Recalling a whole version is the exception and should be treated as that. I just checked some random packages I installed with Scoop and they all use semver. Recalling a whole version would only occur for applications that do not use semver and do not care about versioning at all, which is probably a very small fraction.
It's very common to have two product streams in parallel, like 6.x.x and 7.x.x with pnpm and releasing them both under the releases feature in GitHub. This is actually the case for nearly every application that has LTS or maintenance versions active.
Having a default in Scoop of constantly changing these versions, as they release different ones, is not feasable from my perspective.
I think Scoop should either pin itself to a major version once it was installed or implement the correct updated version checking that I mentioned in the first post, which is already in the code.
